lib/client/api_matrix.es6.js

```diff
@@ -96,35 +114,86 @@ foam.CLASS({
-                    // to ensure we invoke the right function.
-                    if (!Object.prototype.hasOwnProperty.call(matrix[k0], k1)) {
-                      matrix[k0][k1] = {};
-                    }
-                    matrix[k0][k1][k2] = true;
-                  });
-                },
-              },
-            })).then(() => this.filterMatrix_(matrix, options));
+          return new Promise((resolve, reject) => {
+            // TODO(markdittmer): Second argument in sink.put() should be a
+            // subscription, so "detached" here should be unnecessary. Clean
+            // this up once https://github.com/foam-framework/foam2/issues/633
+            // is fixed.
+            let detached = false;
+
+            let promises = [];
+            return this.releaseWebInterfaceJunctionDAO.where(query)
+                .select(this.AnonymousSink.create({
+                  sink: {
+                    put: junction => {
+                      if (detached) return;
+                      if (id && this.matrixId_ !== id) {
+                        detached = true;
+                        resolve(null);
+                        return;
+                      }
+
+                      promises.push(
+                          this.webInterfaceDAO.find(junction.targetId)
+                            .then(iface => {
+                              if (detached) return;
+                              if (id && this.matrixId_ !== id) {
+                                detached = true;
+                                resolve(null);
+                                return;
+                              }
+
+                              const k0 = iface.interfaceName;
+                              const k1 = iface.apiName;
+                              const k2 = junction.sourceId;
+                              // Function and Object's built-in methods can be
+                              // interface name or API name.
+                              // So matrix[k0] is not sufficient to check if
+                              // k0 exits in matrix, use hasOwnProperty instead.
+                              if (!matrix.hasOwnProperty(k0)) {
+                                matrix[k0] = {};
+                              }
+                              // "hasOwnProperty" could also be a apiName,
+                              // thus use Object.prototype.hasOwnProperty
+                              // to ensure we invoke the right function.
+                              if (!Object.prototype.hasOwnProperty.call(
+                                  matrix[k0], k1)) {
+                                matrix[k0][k1] = {};
+                              }
+                              matrix[k0][k1][k2] = true;
+                            }));
+                    },
+                  },
+                })).then(() => {
+                  if (detached) return null;
```

Comment:
I don't know if there's a nice way of fixing it, but these five lines are repeated several times here.

Reply:
I think this should be cleaned up. There are two things that should be done (IMHO):

1. Create a `maybeContinue` that can be `then()` between each step; it will `reject()` to stop flow control (so there will be a `catch()` at the end of the method).
2. We should have a sink that will fetch data in batches and support flow control between batches because the DAOs here are sometimes implemented synchronously (so we'll keep iterating over >1k records synchronously).

Reply:
Updated `TODO` to punt above-mentioned strategy to a future PR.